File: 1. 算法部分/第5章 搜索与回溯算法/12.最佳调度问题/machine.cpp
#include<cstdio>
#include<climits>
#include<cstring>
int n,k,a[101],f[1001],ans;
int max(int x,int y)
{
	if (x>y) return x;
	else return y;
}
void find(int x,int y)
{
	int i;
	if (y>=ans) return;
	if (x>n)
	{
	 if (y<ans) ans=y;
	 return;
	}
	for(i=1;i<=k;i++)
	{
	 f[i]=f[i]+a[x];
	 find(x+1,max(y,f[i]));
	 f[i]=f[i]-a[x];
	}
}
int main()
{
	int i;
	scanf("%d%d",&n,&k);
	for(i=1;i<=n;i++)
	  scanf("%d",&a[i]);
	memset(f,0,sizeof(f));
	ans=INT_MAX;
	find(1,0);
	printf("%d",ans);
}

File: 1. 算法部分/第5章 搜索与回溯算法/12.最佳调度问题/machine.pas
const
        oo=100000000;
var
        n,i,j,k,ans:longint;
        a,l:array[0..21] of longint;
procedure find(x,max:longint);
var
        i:longint;
begin
        if max < ans then
        begin
                if x > n then ans:=max
                else begin
                        for i:=1 to k do
                        begin
                                inc(l[i],a[x]);
                                if l[i] > max then find(x+1,l[i])
                                else find(x+1,max);
                                dec(l[i],a[x]);
                        end;
                     end;
        end;
end;

begin
        assign(input,'machine.in');reset(input);
        assign(output,'machine.out');rewrite(output);
        read(n,k);
        for i:=1 to n do read(a[i]);
        for i:=1 to n-1 do
           for j:=i+1 to n do
              if a[i] < a[j] then
              begin
                a[0]:=a[i];
                a[i]:=a[j];
                a[j]:=a[0];
              end;
        i:=1;
        while I<=n do
        begin
                inc(ans,a[i]);
                i:=i+k;
        end;
        find(1,0);
        writeln(ans);
        close(input);close(output);
end.